yarn-project/p2p/src/service/libp2p_service.ts

```diff
@@ -0,0 +1,385 @@
+      }
+    });
+
+    this.jobQueue.start();
+    await this.node.start();
+    await this.node.handle(this.protocolId, (incoming: IncomingStreamData) =>
+      this.jobQueue.put(() => Promise.resolve(this.handleProtocolDial(incoming))),
+    );
+    const dht = this.node.services['kadDHT'] as DualKadDHT;
+    this.logger(`Started P2P client as ${await dht.getMode()} with Peer ID ${this.node.peerId.toString()}`);
+    this.timeout = setTimeout(async () => {
+      this.logger(`Refreshing routing table...`);
+      await dht.refreshRoutingTable();
+    }, INITIAL_PEER_REFRESH_INTERVAL);
+  }
+
+  /**
+   * Stops the LibP2P service.
+   * @returns An empty promise.
+   */
+  public async stop() {
+    if (this.timeout) {
+      clearTimeout(this.timeout);
+    }
+    await this.jobQueue.end();
+    await this.node.stop();
+  }
+
+  /**
+   * Creates an instance of the LibP2P service.
+   * @param config - The configuration to use when creating the service.
+   * @param txPool - The transaction pool to be accessed by the service.
+   * @returns The new service.
+   */
+  public static async new(config: P2PConfig, txPool: TxPool) {
+    const {
+      enableNat,
+      tcpListenIp,
+      tcpListenPort,
+      announceHostname,
+      announcePort,
+      serverMode,
+      minPeerCount,
+      maxPeerCount,
+    } = config;
+    const peerId = config.peerIdPrivateKey
+      ? await createFromProtobuf(Buffer.from(config.peerIdPrivateKey, 'hex'))
+      : await createLibP2PPeerId();
+
+    const opts: Libp2pOptions<ServiceMap> = {
+      start: false,
+      peerId,
+      addresses: {
+        listen: [`/ip4/${tcpListenIp}/tcp/${tcpListenPort}`],
+        announce: announceHostname ? [`/ip4/${announceHostname}/tcp/${announcePort ?? tcpListenPort}`] : [],
+      },
+      transports: [tcp()],
+      streamMuxers: [yamux(), mplex()],
+      connectionEncryption: [noise()],
+      connectionManager: {
+        minConnections: minPeerCount,
+        maxConnections: maxPeerCount,
+      },
+      peerDiscovery: [
+        bootstrap({
+          list: config.bootstrapNodes,
+        }),
+      ],
+    };
+
+    const services: ServiceFactoryMap = {
+      identify: identifyService({
+        protocolPrefix: 'aztec',
+      }),
+      kadDHT: kadDHT({
+        protocolPrefix: 'aztec',
+        clientMode: !serverMode,
+      }),
+    };
+
+    if (enableNat) {
+      services.nat = autoNATService({
+        protocolPrefix: 'aztec',
+      });
+    }
+
+    const node = await createLibp2p({
+      ...opts,
+      services,
+    });
+    const protocolId = config.transactionProtocol;
+    return new LibP2PService(config, node, protocolId, txPool);
+  }
+
+  /**
+   * Propagates the provided transaction to peers.
+   * @param tx - The transaction to propagate.
+   */
+  public propagateTx(tx: Tx): void {
+    void this.jobQueue.put(() => Promise.resolve(this.sendTxToPeers(tx)));
```

Comment:
Why the Promise.resolve? Is the intention that the queue processor should not await the `sendTxToPeers` to finish before moving onto the next job?

Reply:
That's an error. Nice catch.